### PR TITLE
first draft

### PR DESCRIPTION
**Pull request by scottcreynolds**
@AnnJohn

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,19 +1,111 @@
...
-    * Building for internal use
-      * Discuss the problem of internal apis and auth. Can't store a key on the client, how to do auth?
-      * CORs as a way to stop other domains from using your internal API
-      * Not going to cover auth in a major way in this unit
+## Lesson
+
+We've seen how to *consume* various APIs, now let's talk about
+*providing* one.
+
+The choice to build and provide an API boils down to one of two options:
```
Comment (pletcher) [inline README.md:13]: I really like this framing device -- it helps to focus what can be a kind of unwieldy topic.

File: README.md [hunk 2]
```diff
@@ -1,19 +1,111 @@
...
+
+1. Your data is valuable to more than just your application, or your
+   application has valuable integration points for other
+applications to provide new and interesting services, so you build an *external* API and open it up to third-party developers.
+2. Your data is only valuable to your company, but you access it in
+   multiple ways (e.g. a web application, IOS application, and Android
```
Comment (pletcher) [inline README.md]: Nit: "IOS" -> "iOS"

File: README.md [hunk 3]
```diff
@@ -1,19 +1,111 @@
...
+application all for your company) or have a distributed/service architecture, or you're even just using a front-end framework with Rails as an API back-end, so you build an
```
Comment (pletcher) [inline README.md]: Nits:
- Comma after the close-paren
- "... or your have ... " instead of "... or have ..."

File: README.md [hunk 4]
```diff
@@ -1,19 +1,111 @@
...
+*internal* API to provide common access to all internally-managed
+clients.
+
+### External API Considerations
+
+External APIs are extremely valuable to the growth of the web. Because
+of External APIs you can watch your Uber driver approach and plot your
```
Comment (pletcher) [inline README.md]: Lowercase "External"

File: README.md [hunk 5]
```diff
@@ -1,19 +1,111 @@
...
+course with Google maps and get a text when the driver is close and pay
```
Comment (pletcher) [inline README.md]: Can we replace these "ands" with commas? I feel sort of out of breath by the end of this sentence :)

File: README.md [hunk 6]
```diff
@@ -1,19 +1,111 @@
...
+without ever touching your wallet.
+
+Because an external API is intended to be used by a wide variety of
+developers, with whom you wouldn't have much direct contact, there are a
+lot of things to account for when building one.
+
+**Documentation:** An external API must be *extensively* documented
+   to support your developer community. This includes not only
+functional specifications, but code samples, sometimes in multiple
+languages.
```
Comment (pletcher) [inline README.md]: Spoken or programming languages? (Both, right?)

File: README.md [hunk 7]
```diff
@@ -1,19 +1,111 @@
...
+
+Sometimes you'll want to provide tools to live-test the API, such as with
```
Comment (pletcher) [inline README.md]: Maybe "such as Foursquare does" instead of "such as with Foursquare".

File: README.md [hunk 8]
```diff
@@ -1,19 +1,111 @@
...
+Foursquare. You might even provide libraries (such as a gem) for the
+languages you support.
+
+**Versioning:** We saw this with the Foursquare API. Versioning is an
+   important consideration, and you have to make sure that you don't
+release code in such a way that it suddenly breaks applications that
+rely on your API.
+
+**Authentication and Security:** Allowing external access to your data is a
```
Comment (pletcher) [inline README.md]: Could you add one or two examples of security here to drive the point home? (Maybe just mentioning token-based authentication?)

File: README.md [hunk 9]
```diff
@@ -1,19 +1,111 @@
...
+   serious thing, so your external API needs robust access control and
+security.
+
+**Quality of Service:** Once you have third-party developers relying
+   on your API, there are expectations around quality, uptime, response
```
Comment (pletcher) [inline README.md]: "quality, uptime, and response time" instead of "quality, uptime, response time"

File: README.md [hunk 10]
```diff
@@ -1,19 +1,111 @@
...
+
+**Documentation:** Documentation is still important, but with an
+   internal API you generally won't have to provide support for such a
+wide array of consumers, and the docs can be a little more utilitarian.
+
+**Versioning:** Versioning is still a consideration. Depending on
+   your development process, it's possible that not all of your apps
+will be able to respond to changes at the same time, so you may still
+need versioning to support that. However, the consumers will all be
+known, so you won't have to worry as much about some rogue developer
```
Comment (pletcher) [inline README.md:82]: :D

File: README.md [hunk 11]
```diff
@@ -1,19 +1,111 @@
...
+internal API in cookies or JavaScript, because then it wouldn't be secure. And depending on the
+use of the API, it might not make sense to do user-based authentication
+at all. So how do we ensure that only valid web applications are
+accessing our internal API?
+
+Browsers implement something called [Same Origin Security Policy](https://en.wikipedia.org/wiki/Same-origin_policy), which prevents a page from running a script to access data from a different origin.
+
+This is great because by default it means your API can't be accessed by
+another domain. But what about the ones you want to give access to? For
+that, we can configure [Cross Origin Resource Sharing](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing), or **CORS**, to allow whitelist access to our domains while still blocking others. We won't be doing much with this in this unit, but there are great tools like [rack-cors](https://github.com/cyu/rack-cors) out there to help.
```
Comment (pletcher) [inline README.md]: Trying to put myself in the shoes of a beginning user, I'm a bit confused at the end of this section. How do Same Origin Security Policy and CORS work together to add security? Can we simplify how we describe these tools for users?

For example, even the term "whitelist access" feels a bit jargon-y here. Can we reword this description to make it easier to understand?

File: README.md [hunk 12]
```diff
@@ -1,19 +1,111 @@
...
+internal APIs. In fact, there's a good chance that your internal API is
+mission-critical and handles more requests than an external API might.
+
+You might not need to worry about rate limiting, but peformance, uptime,
+and response time are still extremely important no matter who's
+accessing your API.
+
+## Summary
+
+We've looked at the reasons you might choose to create an internal or
```
Comment (pletcher) [inline README.md:123]: This summary is really helpful. The tone of this lesson is great, it just gets a little dense in the section about security (understandably). Simplifying that section will really tighten the lesson up and help it all cohere.

File: README.md (hunk 5 above)
Comment (scottcreynolds) [inline README.md]: was sorta the idea, narratively, to make it seem intense, but I'll do commas

**Comment by scottcreynolds** [comment]
@pletcher made some edits

File: README.md [hunk 13]
```diff
@@ -1,19 +1,126 @@
...
+release code in such a way that it suddenly breaks applications that
+rely on your API.
+
+**Authentication and Security:** Allowing external access to your data is a
+   serious thing, so your external API needs robust access control and
+security. We came across this when consuming the Foursquare and Github
+APIs. We had to register our application, pass ID and secret pairs to
+the API, and sometimes use OAuth tokens on behalf of a user - all
+security implementations that need to be set up and maintained by the
+API provider.
```
Comment (pletcher) [inline README.md:54]: This is excellent. As a reader, I have a much clearer idea of how to think about authentication and security.

File: README.md [hunk 14]
```diff
@@ -1,19 +1,126 @@
...
+
+**Quality of Service:** Once you have third-party developers relying
+   on your API, there are expectations around quality, uptime, and response
+time.
+
+You also have to consider things like **rate limiting**, where you
+monitor access and only allow a certain number of API calls per
+application in a given timeframe.
+
+We won't be covering much of this in this unit, but these are all
```
Comment (pletcher) [inline README.md]: As a student, when I see the phrase "We won't be covering much of this...", my eyes kind of glaze over and I forget what I'm about to read before even reading it (if that makes any sense).

Maybe we can instead say something that emphasizes that a given implementation of rate limits, uptime, etc. will vary based on the requirements and constraints of the given application, "but these are all important considerations if you plan to make your API public"?

File: README.md [hunk 15]
```diff
@@ -1,19 +1,126 @@
...
+important considerations if you plan to make your API public.
+
+### Internal API Considerations
+
+An internal API is a great way to provide access to data for multiple internal
+applications. While there are still many things to consider, the
+concerns for internal APIs are generally less strenuous because you are
+usually in control of all of the applications that will have access.
```
Comment (pletcher) [inline README.md]: Maybe add "or you can talk to people who are" (parenthetically).

File: README.md [hunk 16]
```diff
@@ -1,19 +1,126 @@
...
+another domain or website.
+
+But what if you want to give another domain access to your API? To do
+that, we can configure [Cross Origin Resource Sharing](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing), or **CORS**. CORS allows us to specify that certain other sites are allowed access while still blocking others.
+
+**Note:** This type of access control is called *whitelisting*. That
+is, creating a list of allowed domains. This is the opposite of
+*blacklist* control, where you allow anything except for a list of
+explicitly blocked sites. Neither should be confused with *The
+Blacklist*, a fun but ridiculous James Spader show.
```
Comment (pletcher) [inline README.md:107]: This great -- I have a much clearer idea of what CORS does/is for.

**Pull request merged**